### ISSUE TITLE
Track range constraints on field elements outside of a circuit

### ISSUE THREAD

**Issue by codygunton** [AztecProtocol/barretenberg#416]
@iAmMichaelConnor  said
> I wonder if it be nice to give a `field_ct` an 'awareness' of its range checks. Similar to `suint_ct`, except it never actually does range checks in its constructor or operators; it just gets "told" its current range by functions around the codebase. `set_current_range`: "you are now definitely less than 32-bits, because I, some function, just checked your range". And then it could track its own crude range upper bound as operations are done to it.

This would enable us to make compile-time assertions about the range of a stdlib field element, and also could be helpful in debugging.

Implementing this is potentially a lot of work. Before building this feature, one should first assess whether this would catch circuit bugs in practice.

**Comment by arielgabizon** [comment]
Definitely the idea of suint was compile time checks.
The idea there was to always have a range constraint in the constructor and never elsehwere.
Mike is suggesting a variant where this one-time range constraint is done not in the constructor but at some later point..
so it would be a variant of suint where the range can be undefined until `set_current_range` is called.
I wonder though if you can't just assume this always happens in the constructor as in suint

**Comment by ledwards2225** [comment]
> suint only does range constraints in the constructor. I don't see how you could say anything about the range if you don't do range constraints anywhere, as in the quoted text of Mike. Definitely the idea of suint was compile time checks.

Yeah what I'm imagining (and I think what Cody meant) was simply some metadata on field that tracks what (if any) range constraints have been applied (both explicitly or implicitly via lookups). Then you could write tests that compare this metadata vs expectation
